File: Lab-1/Task-3/GraphAlgorithmsPackage/src/main/java/com/bibletoon/graphAlgorithms/Algorithms.java
package com.bibletoon.graphAlgorithms;

import java.util.*;

public class Algorithms {
    public static <T> Optional<Node<T>> BFS(T value, Node<T> start) {
        Set<Node<T>> alreadyVisited = new HashSet<Node<T>>();
        Queue<Node<T>> queue = new ArrayDeque<>();
        queue.add(start);

        Node<T> currentNode;

        while (!queue.isEmpty()) {
            currentNode = queue.remove();

            if (currentNode.getValue().equals(value)) {
                return Optional.of(currentNode);
            } else {
                alreadyVisited.add(currentNode);
                queue.addAll(currentNode.getNeighbors());
                queue.removeAll(alreadyVisited);
            }
        }

        return Optional.empty();
    }

    public static <T> Optional<Node<T>> DFS(T value, Node<T> start) {
        Stack<Node<T>> stack = new Stack<>();
        HashSet<Node<T>> alreadyVisited = new HashSet<>();
        stack.push(start);
        while (!stack.isEmpty()) {
            Node<T> current = stack.pop();
            if (!alreadyVisited.contains(current)) {
                alreadyVisited.add(current);
                if (current.getValue() == value)
                    return Optional.of(current);
                for (Node<T> dest : current.getNeighbors()) {
                    if (!alreadyVisited.contains(current))
                        stack.push(dest);
                }
            }
        }

        return Optional.empty();
    }
}

File: Lab-1/Task-3/GraphAlgorithms/GraphAlgorithms/Algorithms.cs
namespace GraphAlgorithms;

public class Algorithms
{
    public static Node<T>? BFS<T>(T value, Node<T> start) {
        HashSet<Node<T>> alreadyVisited = new HashSet<Node<T>>();
        Queue<Node<T>> queue = new Queue<Node<T>>();
        queue.Enqueue(start);

        Node<T> currentNode;

        while (queue.Count != 0) {
            currentNode = queue.Dequeue();

            if (currentNode.getValue().Equals(value)) {
                return currentNode;
            } else {
                alreadyVisited.Add(currentNode);
                foreach (Node<T> neighbor in currentNode.getNeighbors())
                {
                    if (!alreadyVisited.Contains(neighbor))
                        queue.Enqueue(neighbor);
                }
            }
        }

        return null;
    }

    public static Node<T>? DFS<T>(T value, Node<T> start) {
        Stack<Node<T>> stack = new Stack<Node<T>>();
        HashSet<Node<T>> alreadyVisited = new HashSet<Node<T>>();
        stack.Push(start);
        while (stack.Count != 0) {
            Node<T> current = stack.Pop();
            if (!alreadyVisited.Contains(current)) {
                alreadyVisited.Add(current);

                if (current.getValue().Equals(value))
                    return current;
                foreach (Node<T> dest in current.getNeighbors()) 
                {
                    if (!alreadyVisited.Contains(current))
                        stack.Push(dest);
                }
            }
        }

        return null;
    }
}